server/sonar-server/src/main/java/org/sonar/server/computation/taskprocessor/CeTaskProcessorRepositoryImpl.java

```diff
@@ -0,0 +1,104 @@
+
+import com.google.common.base.Function;
+import com.google.common.base.Joiner;
+import com.google.common.base.Optional;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Multimap;
+import java.util.Collection;
+import java.util.Map;
+import javax.annotation.Nonnull;
+import org.sonar.server.computation.queue.CeTask;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.FluentIterable.from;
+import static java.lang.String.CASE_INSENSITIVE_ORDER;
+import static java.lang.String.format;
+
+/**
+ * {@link CeTaskProcessorRepository} implementation which provides access to the {@link CeTaskProcessor} existing in the
+ * PicoContainer the current object belongs to.
+ */
+public class CeTaskProcessorRepositoryImpl implements CeTaskProcessorRepository {
+  private static final Joiner COMMA_JOINER = Joiner.on(", ");
+
+  private final Map<String, CeTaskProcessor> taskProcessorByCeTaskType;
+
+  public CeTaskProcessorRepositoryImpl(CeTaskProcessor[] taskProcessors) {
+    this.taskProcessorByCeTaskType = indexTaskProcessors(taskProcessors);
+  }
+
+  @Override
+  public Optional<CeTaskProcessor> getForCeTask(CeTask ceTask) {
+    return Optional.fromNullable(taskProcessorByCeTaskType.get(ceTask.getType()));
+  }
+
+  private static Map<String, CeTaskProcessor> indexTaskProcessors(CeTaskProcessor[] taskProcessors) {
+    Multimap<String, CeTaskProcessor> permissiveIndex = buildPermissiveCeTaskProcessorIndex(taskProcessors);
+    checkUniqueHandlerPerCeTaskType(permissiveIndex);
+    return ImmutableMap.copyOf(Maps.transformValues(permissiveIndex.asMap(), CeTaskProcessorCollectionToFirstElement.INSTANCE));
+  }
+
+  private static Multimap<String, CeTaskProcessor> buildPermissiveCeTaskProcessorIndex(CeTaskProcessor[] taskProcessors) {
+    Multimap<String, CeTaskProcessor> permissiveIndex = ArrayListMultimap.create(taskProcessors.length, 1);
+    for (CeTaskProcessor taskProcessor : taskProcessors) {
+      for (String ceTaskType : taskProcessor.getHandledCeTaskTypes()) {
+        permissiveIndex.put(ceTaskType, taskProcessor);
+      }
+    }
+    return permissiveIndex;
+  }
+
+  private static void checkUniqueHandlerPerCeTaskType(Multimap<String, CeTaskProcessor> permissiveIndex) {
+    for (Map.Entry<String, Collection<CeTaskProcessor>> entry : permissiveIndex.asMap().entrySet()) {
+      checkArgument(
+        entry.getValue().size() == 1,
+        format(
+          "There can be only one CeTaskProcessor instance registered as the processor for CeTask type %s. " +
+            "More than one found. Please fix your configuration: %s",
+          entry.getKey(),
+          COMMA_JOINER.join(from(entry.getValue()).transform(ToClassName.INSTANCE).toSortedList(CASE_INSENSITIVE_ORDER))));
+    }
+  }
+
+  private enum ToClassName implements Function<Object, String> {
```

Comment:
We already discussed in the past of a util class to share all these common functions. To keep in mind for later.